hivemq-edge/src/main/java/com/hivemq/extensions/core/CommercialModuleLoaderDiscovery.java

```diff
@@ -15,40 +15,73 @@
  */
 package com.hivemq.extensions.core;
 
+import com.hivemq.bootstrap.services.CompleteBootstrapService;
+import com.hivemq.bootstrap.services.GeneralBootstrapService;
+import com.hivemq.bootstrap.services.PersistenceBootstrapService;
 import com.hivemq.edge.modules.ModuleLoader;
 import com.hivemq.extension.sdk.api.annotations.NotNull;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
 import java.util.List;
 
 public class CommercialModuleLoaderDiscovery {
 
     private static final @NotNull Logger log = LoggerFactory.getLogger(CommercialModuleLoaderDiscovery.class);
     private final @NotNull ModuleLoader moduleLoader;
-    private final @NotNull CoreModuleServiceImpl coreModuleService;
+    private final List<ModuleLoaderMain> instances = new ArrayList<>();
 
     public CommercialModuleLoaderDiscovery(
-            final @NotNull ModuleLoader moduleLoader, final @NotNull CoreModuleServiceImpl coreModuleService) {
+            final @NotNull ModuleLoader moduleLoader) {
         this.moduleLoader = moduleLoader;
-        this.coreModuleService = coreModuleService;
     }
 
-    public void loadAllCoreModules()
-            throws InvocationTargetException, InstantiationException, IllegalAccessException, NoSuchMethodException {
+    public void discoverModuleLoaderMainClasses() {
         moduleLoader.loadModules();
         final List<Class<? extends ModuleLoaderMain>> implementations =
                 moduleLoader.findImplementations(ModuleLoaderMain.class);
-        for (Class<? extends ModuleLoaderMain> implementation : implementations) {
-            loadAndStartMainClass(implementation);
+
+        implementations.forEach(impl -> {
+            try {
+                this.instances.add(impl.getDeclaredConstructor().newInstance());
+            } catch (Exception e) {
+                log.error("Error when instancing '{}':", impl, e);
+            }
+        });
+    }
+
+    public void generalBootstrap(final @NotNull GeneralBootstrapService generalBootstrapService) {
+        try {
+            for (ModuleLoaderMain instance : instances) {
+                instance.generalBootstrap(generalBootstrapService);
+            }
+        } catch (Exception e) {
+            log.error("Error when bootstrapping general information", e);
+        }
+    }
+
+    public void persistenceBootstrap(final @NotNull PersistenceBootstrapService persistenceBootstrapService) {
+        try {
+            for (ModuleLoaderMain instance : instances) {
```

Comment:
discoverModuleLoaderMainClasses can be triggered while iterating which would lead to a. ConcurrentModificationException.
Either iterate using an iterator or an index-based for takes care of that.

Reply:
yeah if the method would be called a second time, that is true. It is supposed to be called once before bootstraping everything. So right now it is safe, however for the future you are right. If somebody calls it a second time and concurrently, we get that exception. Index based is not safe as the List could get shorter though (theoretically).